[LTC] Upstream utils in computation_client

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69815
* #69716
* __->__ #69621

Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

desertfire committed:
Upstream utils in computation_client

Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

desertfire committed:
Update on "Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire committed:
Update on "Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

desertfire commented:
@JackCaoG , is pytorch/xla already using code from LTC? I wonder how this PR can trigger a XLA test failure?

JackCaoG commented:
@desertfire It is a pt/xla head issue, I just disabled the test and will look into it a bit more. Don't worry about it for this pr

alanwaketan commented on torch/csrc/lazy/core/config.h:
```diff
@@ -1,4 +1,7 @@
 #pragma once
 #include <c10/util/Flags.h>
 
+C10_DECLARE_int(torch_lazy_io_thread_pool_size);
```
I will let @wconstab comment on these.

desertfire replied:
Do you mean how to setup them through environment variables?

wconstab replied:
I'm not sure what @alanwaketan means.

I made a bad assumption that all the declared flags would automatically be available as env vars or cmd line options.  I think this is true for gflags in some scenarios, but, i couldn't figure out how to set the env vars in torch so far.

So ultimately, i think we'll have to
 - convert some of these to pybindings and decide they don't need env vars at all
 - for ones that need env vars, somehow add support - either switch back to getenv, or, figure out how to turn on env parsing for C10_DECLARE flags.
 
 It probably doesn't hurt to keep following this pattern of moving all the envs into one config file though.  Might make it easier later on to make the above changes all-at-once?

alanwaketan reviewed (approved):
Otherwise, LGTM.

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire committed:
Update on "Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32957629).

desertfire committed:
Update on "[LTC] Upstream utils in computation_client"


Summary: Upstream the following utils
- metrics.h
- multi_wait.h
- thread_pool.h
- unique.h

Differential Revision: [D32957629](https://our.internmc.facebook.com/intern/diff/D32957629)

[ghstack-poisoned]
